### DISCUSSION TITLE
Abbreviation or renaming of layout controls

### DISCUSSION THREAD

**Discussion by sacdevops** [dotnet/maui#4421] [Ideas]
Based on the issue [GridLayout -> Grid to reduce confusion](https://github.com/dotnet/maui/pull/3865) all layout controls should be renamed. 

My proposition: 

- AbsoluteLayout -> Absolute
- RelativeLayout -> Relative
- FlexLayout -> Flex
- StackLayout -> Stack
- HorizontalStackLayout -> HStack
- VerticalStackLayout -> VStack

The abbreviation of the terms speaks for itself. It's ultimately a matter of taste, but shorter terms lead to a cleaner and faster typable variant.

This would be a wish before Maui is made available for release as it is again a code change for .NET Maui users. Such decisions should be implemented in advance. It is also not a big effort to change the code status. It is important that we decide on one variant right from the start.

What is your opinion on this?

**Comment by saint4eva** [comment]
I think Comet is already using VStack and HStack, so another name should be chosen for Maui to avoid collision.

**Reply by sacdevops** [reply]
Maybe Comet should change it then 😄 I prefer Row and Column like in Flutter, but it would not suit with the StackLayout that will be still a part of Maui.

**Reply by saint4eva** [reply]
Why should Comet change?

**Comment by hartez** [comment]
> Based on the issue [GridLayout -> Grid to reduce confusion](https://github.com/dotnet/maui/pull/3865) all layout controls should be renamed.
> 
> My proposition:
> 
> * AbsoluteLayout -> Absolute
> * RelativeLayout -> Relative
> * FlexLayout -> Flex

As I said in the dicussion in the issue linked above, "Absolute", "Relative", and "Flex" are precisely the reason we will _not_ be doing this. Those names are not descriptive for new users (they give no indication that they have anything to do with layout). 

> * StackLayout -> Stack

Stack" is even more problematic, because it conflicts with the name of a common data structure and an existing .NET type. 

> * HorizontalStackLayout -> HStack
> * VerticalStackLayout -> VStack
> 
> The abbreviation of the terms speaks for itself. It's ultimately a matter of taste, but shorter terms lead to a cleaner and faster typable variant.

"HStack" and "VStack" violate the [framework naming conventions](https://docs.microsoft.com/en-us/dotnet/standard/design-guidelines/general-naming-conventions), so for consistency we will not be using those names for MAUI Controls. 

If being able to type them faster is a concern, you can always subclass them (`public class HStack : HorizontalStackLayout`). 🙂

> It's ultimately a matter of taste

That's one concern; another is portability from Forms. As much as I'd prefer consistent naming for aesthetic purposes, we also have to balance that with the fact that existing users need to update their apps to MAUI.

**Reply by sacdevops** [reply]
>  As I said in the dicussion in the issue linked above, "Absolute", "Relative", and "Flex" are precisely the reason we will not be doing this. 
Those names are not descriptive for new users (they give no indication that they have anything to do with layout).

I disagree. Flex is a familiar term that most HTML users should know by now. Sure, it's called flexbox, but flex is still a well-known term. The additional layout is unnecessary here.
I agree with you about the terms relative and absolute, but these terms are poorly chosen from the ground up. 

> "HStack" and "VStack" violate the framework naming conventions, so for consistency we will not be using those names for MAUI Controls.

Rules are there to be broken. The Maui team should then think about how to improve the whole thing.
People always say that it should be an alternative to other technologies. But we don't want an alternative, we want reasonable competition. Currently, Maui is completely bypassing wishful thinking and is becoming more and more like an XF6 by the day. You are in the process of revising the basic foundation. There is a saying from German:
Do one thing carefully and slowly, but only do it once.
You're not going to completely overturn the project and redo it again.
Just do it right for fun. 
 
> If being able to type them faster is a concern, you can always subclass them (public class HStack : HorizontalStackLayout).

Why do we always have to work with dirty solutions here? You can really sit down at the table and come up with a sensible solution instead of always demanding everything from the community. 

> That's one concern; another is portability from Forms. As much as I'd prefer consistent naming for aesthetic purposes, we also have to balance that with the fact that existing users need to update their apps to MAUI.

That's why I'm talking about it. As a community, you should agree on a name. Of course, not everyone will like it, but the topic should be addressed and approved by the majority of the community. 
The team or Microsoft won't be using Maui anyway. This is just a tool for us. Then we should also feel comfortable here and express our wishes.

**Reply by osnipezzini** [reply]
> > Based on the issue [GridLayout -> Grid to reduce confusion](https://github.com/dotnet/maui/pull/3865) all layout controls should be renamed.
> > My proposition:
> > 
> > * AbsoluteLayout -> Absolute
> > * RelativeLayout -> Relative
> > * FlexLayout -> Flex
> 
> As I said in the dicussion in the issue linked above, "Absolute", "Relative", and "Flex" are precisely the reason we will _not_ be doing this. Those names are not descriptive for new users (they give no indication that they have anything to do with layout).
> 
> > * StackLayout -> Stack
> 
> Stack" is even more problematic, because it conflicts with the name of a common data structure and an existing .NET type.
> 
> > * HorizontalStackLayout -> HStack
> > * VerticalStackLayout -> VStack
> > 
> > The abbreviation of the terms speaks for itself. It's ultimately a matter of taste, but shorter terms lead to a cleaner and faster typable variant.
> 
> "HStack" and "VStack" violate the [framework naming conventions](https://docs.microsoft.com/en-us/dotnet/standard/design-guidelines/general-naming-conventions), so for consistency we will not be using those names for MAUI Controls.
> 
> If being able to type them faster is a concern, you can always subclass them (`public class HStack : HorizontalStackLayout`). 🙂
> 
> > It's ultimately a matter of taste
> 
> That's one concern; another is portability from Forms. As much as I'd prefer consistent naming for aesthetic purposes, we also have to balance that with the fact that existing users need to update their apps to MAUI.

I totally agree with that, removing the layout suffix will cause confusion among newer ones. And also, why bother with it??

**Reply by saint4eva** [reply]
> . The team or Microsoft won't be using Maui anyway. This is just a tool for us. Then we should also feel comfortable here and express our wishes.

Some comments baffle one sometimes.
